services/statestorage/service.go

```diff
@@ -1,44 +1,74 @@
 package statestorage
 
 import (
+	"fmt"
 	"github.com/orbs-network/orbs-network-go/services/statestorage/adapter"
 	"github.com/orbs-network/orbs-spec/types/go/protocol"
 	"github.com/orbs-network/orbs-spec/types/go/services"
 )
 
 type service struct {
-	services.StateStorage
-	persistence adapter.StatePersistence
+	persistence            adapter.StatePersistence
+	lastResultsBlockHeader *protocol.ResultsBlockHeader
 }
 
 func NewStateStorage(persistence adapter.StatePersistence) services.StateStorage {
 	return &service{
-		persistence: persistence,
+		persistence:            persistence,
+		lastResultsBlockHeader: (&protocol.ResultsBlockHeaderBuilder{}).Build(),
 	}
 }
 
 func (s *service) CommitStateDiff(input *services.CommitStateDiffInput) (*services.CommitStateDiffOutput, error) {
+	committedBlock := input.ResultsBlockHeader.BlockHeight()
+	if lastCommittedBlock := s.lastResultsBlockHeader.BlockHeight(); lastCommittedBlock+1 != committedBlock {
+		return &services.CommitStateDiffOutput{NextDesiredBlockHeight: lastCommittedBlock + 1}, nil
+	}
+
 	for _, stateDiffs := range input.ContractStateDiffs {
 		for i := stateDiffs.StateDiffsIterator(); i.HasNext(); {
-			s.persistence.WriteState(i.NextStateDiffs())
+			s.persistence.WriteState(stateDiffs.ContractName(), i.NextStateDiffs())
 		}
 	}
-
-	return nil, nil
+	s.lastResultsBlockHeader = input.ResultsBlockHeader
+	hi := committedBlock + 1
+	return &services.CommitStateDiffOutput{NextDesiredBlockHeight: hi}, nil
 }
 
 func (s *service) ReadKeys(input *services.ReadKeysInput) (*services.ReadKeysOutput, error) {
-	var state []*protocol.StateRecord
-	for _, stateDiff := range s.persistence.ReadState() {
-		state = append(state, stateDiff)
+	if input.ContractName == "" {
+		return nil, fmt.Errorf("missing contract name")
+	}
+
+	contractState := s.persistence.ReadState(input.ContractName)
```

Comment:
I think it would be better if ReadState would return err. Than you would avoid the if contractState == nil  and just check the err